File: app/src/main/java/org/d1scw0rld/bookbag/dto/Changeable.java
package org.d1scw0rld.bookbag.dto;

import android.support.annotation.NonNull;

public class Changeable<T>
{
   public T value;

   Changeable(T value)
   {
      this.value = value;
   }

   @NonNull
   public String toString() 
   {
      return value.toString();
   }
   
   public Class getGenericType()
   {
      return value.getClass();
   }
   
//   public void set(T value)
//   {
//      this.value = value;
//   }

   public boolean equals(Object other) 
   {
      if (other instanceof Changeable) 
         return value.equals(((Changeable)other).value);
      else 
         return value.equals(other);
   }

   public int hashCode() 
   {
      return value.hashCode();
   }
   
   public boolean isEmpty()
   {
      if(value instanceof String)
         return ((String) value).trim().isEmpty();
      else if(value instanceof Integer)
         return ((Integer)value) == 0;
      return true;
   }
}

File: app/src/main/java/org/d1scw0rld/bookbag/dto/Changeable.kt
package org.d1scw0rld.bookbag.dto

class Changeable<T : Any>(@JvmField var value: T) {

    override fun toString(): String {
        return value.toString()
    }

    /**
     * Returns the runtime class of the wrapped value.
     * Note: Rename from [getGenericType] to [valueType] reflects that generic type information 
     * is erased at runtime and this returns the concrete class of the [value].
     */
    val valueType: Class<out Any>
        get() = value.javaClass

    @Deprecated(
        message = "Use valueType instead, as this returns the concrete runtime class rather than the compile-time generic type parameter.",
        replaceWith = ReplaceWith("valueType")
    )
    fun getGenericType(): Class<out Any> {
        return valueType
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Changeable<*>) return false
        return value == other.value
    }

    override fun hashCode(): Int {
        return value.hashCode()
    }

    fun isEmpty(): Boolean {
        return when (val v = value) {
            is CharSequence -> v.trim().isEmpty()
            is Number -> v.toInt() == 0
            else -> false // Safe default for arbitrary objects which are non-null
        }
    }
}
